Issue #92 Refactoring of goroutine handling and conditions

hferentschik opened the pull request:
- Refacotring goroutines to all be controlled and started from main
- Creating one UserIdler instance and goroutine per usern/namepace to watch, instead
  of looping all users in a tight for loop
- Making sure goroutines are canelable
- Refactoring conditions

Addresses issues #65, #92 and #99.

hferentschik commented:
@vpavlin FYI, I am about to change the behaviour of _OpenShiftController.Run_, so that users are not checked in a tight for loop. Work in progress...

vpavlin commented on internal/condition/deployment_config_condition.go:
```diff
@@ -0,0 +1,35 @@
+
+// DCCondition covers changes to Jenkins DeploymentConfigs
+type DeploymentConfigCondition struct {
+	Condition
```
`Condition` is referenced here, but not in other conditions, is that intentional?

hferentschik replied:
ahh, my wrong. I thought I removed it. I pushed an update ...

vpavlin commented:
I went through it and it generally looks good. I'll try in IDE tomorrow so that I can jum around a bit easier:)

One question though - if I understand correctly it now

* Creates a channel per user
* Creates a goroutine which handles that user

Is that correct? 

If yes, and sorry for my dumb question, how will golang cope with, let's say, 3000 channels and goroutines like this? (I don't remember the numbers, but assume that's potentially what we can get on the clusters?)

hferentschik commented:
> I went through it and it generally looks good. I'll try in IDE tomorrow so that I can jump around a bit easier:)

Please do. I did some more changes right now, so make sure to get the latest in case you pulled it already down. As said, there are still a couple of todos and I fully expect that there are still some mistakes in the code. I was concerned about basic flow and code organization so far. My plan is to add some more tests now and also taking the idler for some test spins now. Hopefully tomorrow, but I fear there will be a whole bunch of meetings as well as we need to drive the migration tasks forward for the proxy. 

> If yes, and sorry for my dumb question, how will golang cope with, let's say, 3000 channels and goroutines like this? 

He he, a natural concern. As old Java programmer, I was weary as well. However, my understanding is that in Go creating these many goroutines is no issue at all and in fact expected. I am basing this on my reading of "Concurrency in Go" out of which I also took some of the patterns you see in the code. See also https://stackoverflow.com/questions/8509152/max-number-of-goroutines. 

> It is practical to create hundreds of thousands of goroutines in the same address space.

In this context, I am quite confident that we run first out of cluster space than the ability to create and maintain the goroutines.

Architectually it is quite neat imo. Not only do we get rid of tight for loop iterating over the known users, but we also make sure that processing of an event for a given user is not "blocked" by other users which are processed before this user.

But of course we will have to have a close eye on this.

hferentschik commented:
> I'll try in IDE tomorrow so that I can jump around a bit easier

BTW, have you tried GoLand from Intellij. Until recently I was using Idea with the Go plugin which was already pretty good, but now Intellij released a Go specific version of their IDE - GoLand. Absolutely awesome. Beats imo any other IDE by miles.

vpavlin commented:
> In this context, I am quite confident that we run first out of cluster space than the ability to create and maintain the goroutines.

:+1: 

> Architectually it is quite neat imo

I agree, it looks really good and clean


> GoLand from Intellij.

No, I use VS Code with some golang plugin

hferentschik commented:
@vpavlin fyi, I am back on this one now. I added a commit to also run the show locally. Also removed dependency to Jenkins Proxy. Work in progress ...

vpavlin commented on scripts/setupLocalIdler.sh:
```diff
@@ -0,0 +1,211 @@
+Usage: ${0##*/} [start|stop]
+
+This script is used to run the Jenkins Idler on localhost.
+As a prerequisite DSAAS_PREVIEW_TOKEN, OPENSHIFT_API_TOKEN and  JC_AUTH_TOKEN need to be exported
```
```
s/OPENSHIFT_API_TOKEN/JC_OPENSHIFT_API_TOKEN/
```

vpavlin commented on scripts/setupLocalIdler.sh:
```diff
@@ -0,0 +1,211 @@
+As a prerequisite DSAAS_PREVIEW_TOKEN, OPENSHIFT_API_TOKEN and  JC_AUTH_TOKEN need to be exported
+in the shell you run this script.
+
+DSAAS_PREVIEW_TOKEN is the OpenShift token for console.rh-idev.openshift.com, OPENSHIFT_API_TOKEN is
```
```
s/OPENSHIFT_API_TOKEN/JC_OPENSHIFT_API_TOKEN/
```

vpavlin commented:
I've pushed a commit to change the local dev script slightly - it uses the namespace and token directly without login - I think that is nicer experience (as people might be logged into different cluster and after using the script, they need to re-login)

Otherwise, it seemed to "almost" work - I could not get unleash to return true for my user:/, so I'd say let's merge and see how it behaves in the cluster

vpavlin commented on internal/condition/user_condition.go:
```diff
@@ -0,0 +1,86 @@
+// Eval returns true if there are no buffered request, the last forwarded request occurred more than UserCondition.idleAfter
+// minutes ago and the user accessed the Jenkins UI more than UserCondition.idleAfter minutes ago.
+func (c *UserCondition) Eval(object interface{}) (bool, error) {
+	b, ok := object.(*model.User)
```
This should not be pointer to User, but rather just user (other conditions are ok)

```
{"level":"error","msg":"{{{  {    } %!s(int=0)} {New 0001-01-01 00:00:00 +0000 UTC 0001-01-01 00:00:00 +0000 UTC} {%!s(int=0) {}}} {{  {    } %!s(int=0)} { 0001-01-01 00:00:00 +0000 UTC 0001-01-01 00:00:00 +0000 UTC} {%!s(int=0) {}}} stg2-ebaron1 [{%!s(bool=false) 2018-02-15 14:35:34.034718926 +0000 UTC init} {%!s(bool=true) 2018-02-15 14:36:34.281550291 +0000 UTC Jenkins Unidled for  at 0001-01-01 00:00:00 +0000 UTC}] %!s(int=0) %!s(int=1) 0001-01-01 00:00:00 +0000 UTC 6caa5e71-7c95-40a3-aeb6-cfbd75bf8921} is not of type *User","time":"2018-02-15T15:56:37+01:00"}
```

at the end **is not of type \*User**

hferentschik replied:
ahh, good catch

vpavlin commented on scripts/setupLocalIdler.sh:
```diff
@@ -0,0 +1,222 @@
+    echo export JC_IDLE_AFTER=${LOCAL_DEFAULT_IDLE_TIME}
+    echo export JC_JENKINS_PROXY_API_URL=http://localhost:${LOCAL_PROXY_PORT}
+    echo export JC_F8TENANT_API_URL=http://localhost:${LOCAL_TENANT_PORT}
+    echo export JC_TOGGLE_API_URL=http://localhost:${LOCAL_TOGGLE_PORT}
```
Add `/api` to the toggle url

hferentschik replied:
doh

hferentschik replied:
now I get enabled users as well. nice

hferentschik commented on scripts/setupLocalIdler.sh:
```diff
@@ -44,6 +44,19 @@ To stop:
+# Returns:
+#   None
+###############################################################################
+loc() {
```
ahh, interesting. That's neat

hferentschik replied:
I'll integrate this.

hferentschik commented:
@vpavlin, I removed the WIP status. After adding your fixes, there was just a couple of minor thing.s Now things behave as they should afaict. At least all errors are gone, I can see enabled users and I also can trigger a build and see it picked up and handled by the Idler.

I also can now manually Idle and it won't un-idle until the next timeout which was one of the initial goals.

Really some more tests would be nice. As you might recall the whole point of the exercise was to be able to also test the proxy functionality with actual idle/un-idle. So can do that now :-)

That said, given the current situation, I think it makes sense to get this in asap. It is definitely a step forward and it might anyway be better to take any upcoming things as new issues.

Let's keep the commits intact. I think it makes sene to have mutliple commits and they are actually nicely separated now, each standing pretty much by itself.
